utxonursery.go

```diff
@@ -505,426 +516,855 @@ func (u *utxoNursery) NurseryReport(chanPoint *wire.OutPoint) (*contractMaturity
+		&finalTxID, u.cfg.ConfDepth, heightHint)
+	if err != nil {
+		utxnLog.Errorf("unable to register notification for "+
+			"sweep confirmation: %v", finalTxID)
+		return err
+	}
+
+	utxnLog.Infof("Registering sweep tx %v for confs at height=%d",
+		finalTxID, heightHint)
+
+	u.wg.Add(1)
+	go u.waitForSweepConf(heightHint, kgtnOutputs, confChan)
+
+	return nil
+}
+
+// waitForSweepConf watches for the confirmation of a sweep transaction
+// containing a batch of kindergarten outputs. Once confirmation has been
+// received, the nursery will mark those outputs as fully graduated, and proceed
+// to mark any mature channels as fully closed in channeldb.
+// NOTE(conner): this method MUST be called as a go routine.
+func (u *utxoNursery) waitForSweepConf(classHeight uint32,
+	kgtnOutputs []kidOutput, confChan *chainntnfs.ConfirmationEvent) {
+
+	defer u.wg.Done()
+
+	select {
+	case _, ok := <-confChan.Confirmed:
+		if !ok {
+			utxnLog.Errorf("Notification chan closed, can't"+
+				" advance %v graduating outputs",
+				len(kgtnOutputs))
+			return
 		}
 
-		sweptOutputs, err := deserializeKidList(bytes.NewBuffer(results))
-		if err != nil {
-			return err
+	case <-u.quit:
+		return
+	}
+
+	// TODO(conner): add retry logic?
+
+	u.mu.Lock()
+	defer u.mu.Unlock()
+
+	// Mark the confirmed kindergarten outputs as graduated.
+	if err := u.cfg.Store.GraduateKinder(classHeight); err != nil {
+		utxnLog.Errorf("Unable to graduate %v kingdergarten outputs: "+
+			"%v", len(kgtnOutputs), err)
+		return
+	}
+
+	utxnLog.Infof("Graduated %d kindergarten outputs from height=%d",
+		len(kgtnOutputs), classHeight)
+
+	// Iterate over the kid outputs and construct a set of all channel
+	// points to which they belong.
+	var possibleCloses = make(map[wire.OutPoint]struct{})
+	for _, kid := range kgtnOutputs {
+		possibleCloses[*kid.OriginChanPoint()] = struct{}{}
+
+	}
+
+	// Attempt to close each channel, only doing so if all of the channel's
+	// outputs have been graduated.
+	for chanPoint := range possibleCloses {
+		if err := u.closeAndRemoveIfMature(&chanPoint); err != nil {
+			utxnLog.Errorf("Failed to close and remove channel %v",
+				chanPoint)
+			return
 		}
-		utxnLog.Infof("Deleting %v swept outputs from kindergarten bucket "+
-			"at block height: %v", len(sweptOutputs), deleteHeight)
-
-		// Additionally, for each output that has now been fully swept,
-		// we'll also remove the index entry for that output.
-		indexBucket := tx.Bucket(contractIndex)
-		if indexBucket == nil {
-			return nil
+	}
+}
+
+// sweepCribOutput broadcasts the crib output's htlc timeout txn, and sets up a
+// notification that will advance it to the kindergarten bucket upon
+// confirmation.
+func (u *utxoNursery) sweepCribOutput(classHeight uint32, baby *babyOutput) error {
+	// Broadcast HTLC transaction
+	// TODO(conner): handle concrete error types returned from publication
+	err := u.cfg.PublishTransaction(baby.timeoutTx)
```

Comment:
Can you add a log message here dumping the full second-level transaction along with some of the details within the `babyOutput`? Thanks!

Reply:
of course!